### ISSUE TITLE
Question: are we not cheating by performing (by now thousands or tens of thousands) of architecture and hyperparameter iterations over the same val set?

### ISSUE THREAD

**Issue by Any-Winter-4079** [KellerJordan/modded-nanogpt#223]
Hey everyone,

Title, basically. So, while improvements can be statistically significant for the validation set, they might not be for other 10M token chunks.

I know this might not the purpose of this repo, but if we hope to paint a more accurate picture of the generalization of these techniques, aren't we tricking ourselves (a little bit at least) by thinking we are reducing training time so much?
Has someone tested performance on other val chunks, to see if we are overfitting to **this** val set?

**Comment by ClassicLarry** [comment]
It would be interesting to see someone quantify this over randomly sampled validation sets.

**Comment by ClassicLarry** [comment]
I tested this on 30 unique validation sets, after adding steps worth 10s to the runtime.

Based on the data below, I find it plausible that the improvements from 45 minutes to 90 seconds includes roughly 5 seconds of overfitting, and at least 1 second. Which, all things considered, I find surprisingly low.

There is a natural 5-15s variance in speed required to achieve 3.28 loss between each validation set. The first line gets the lowest loss, which is the one we test on. Quite a few of the 10 million token sets would probably have the record around 91s instead of 90s, which are sets that no one has ever looked at in modded-nanogpt. Some of the really unlucky validation sets would have the record up around 105-110s.

step:1730/1730 val_loss:3.2539 train_time:101622ms step_avg:58.74ms
step:1730/1730 val_loss:3.2878 train_time:101622ms step_avg:58.74ms
step:1730/1730 val_loss:3.2760 train_time:101622ms step_avg:58.74ms
step:1730/1730 val_loss:3.2779 train_time:101622ms step_avg:58.74ms
step:1730/1730 val_loss:3.2681 train_time:101622ms step_avg:58.74ms
step:1730/1730 val_loss:3.2809 train_time:101622ms step_avg:58.74ms
step:1730/1730 val_loss:3.2598 train_time:101622ms step_avg:58.74ms
step:1730/1730 val_loss:3.2654 train_time:101622ms step_avg:58.74ms
step:1730/1730 val_loss:3.2544 train_time:101622ms step_avg:58.74ms
step:1730/1730 val_loss:3.2878 train_time:101622ms step_avg:58.74ms
step:1730/1730 val_loss:3.2772 train_time:101622ms step_avg:58.74ms
step:1730/1730 val_loss:3.3021 train_time:101622ms step_avg:58.74ms
step:1730/1730 val_loss:3.2951 train_time:101622ms step_avg:58.74ms
step:1730/1730 val_loss:3.3038 train_time:101622ms step_avg:58.74ms
step:1730/1730 val_loss:3.2778 train_time:101622ms step_avg:58.74ms
step:1730/1730 val_loss:3.2837 train_time:101622ms step_avg:58.74ms
step:1730/1730 val_loss:3.2806 train_time:101622ms step_avg:58.74ms
step:1730/1730 val_loss:3.2763 train_time:101622ms step_avg:58.74ms
step:1730/1730 val_loss:3.2570 train_time:101622ms step_avg:58.74ms
step:1730/1730 val_loss:3.2636 train_time:101622ms step_avg:58.74ms
step:1730/1730 val_loss:3.2709 train_time:101622ms step_avg:58.74ms
step:1730/1730 val_loss:3.2729 train_time:101622ms step_avg:58.74ms
step:1730/1730 val_loss:3.2546 train_time:101622ms step_avg:58.74ms
step:1730/1730 val_loss:3.2704 train_time:101622ms step_avg:58.74ms
step:1730/1730 val_loss:3.2666 train_time:101622ms step_avg:58.74ms
step:1730/1730 val_loss:3.2681 train_time:101622ms step_avg:58.74ms
step:1730/1730 val_loss:3.2758 train_time:101622ms step_avg:58.74ms
step:1730/1730 val_loss:3.2749 train_time:101622ms step_avg:58.74ms
step:1730/1730 val_loss:3.2850 train_time:101622ms step_avg:58.74ms
step:1730/1730 val_loss:3.2741 train_time:101622ms step_avg:58.74ms
step:1730/1730 val_loss:3.2630 train_time:101622ms step_avg:58.74ms

**Comment by ClassicLarry** [comment]
the test above is not very scientific, as I don't control for the difficulty of each batch by evaluating a naive baseline transformer on each validation set. Would be cool if somebody tests this in a more controlled way!
